controller/scheduler/job.go

```diff
@@ -0,0 +1,181 @@
+package main
+
+import (
+	"errors"
+	"time"
+
+	ct "github.com/flynn/flynn/controller/types"
+	"github.com/flynn/flynn/controller/utils"
+	"github.com/flynn/flynn/host/types"
+)
+
+type JobRequestType string
+type JobState string
+
+const (
+	JobRequestTypeUp   JobRequestType = "up"
+	JobRequestTypeDown JobRequestType = "down"
+	JobStateStarting   JobState       = "starting"
+	JobStateRunning    JobState       = "running"
+	JobStateStopping   JobState       = "stopping"
+	JobStateStopped    JobState       = "stopped"
+	JobStateCrashed    JobState       = "crashed"
+	JobStateRequesting JobState       = "requesting"
+	JobStateNew        JobState       = "new"
+)
+
+type JobRequest struct {
+	*Job
+	RequestType JobRequestType
+	attempts    uint
+}
+
+func NewJobRequest(f *Formation, requestType JobRequestType, typ, hostID, jobID string) *JobRequest {
+	return &JobRequest{
+		Job:         NewJob(f, f.App.ID, f.Release.ID, typ, hostID, jobID),
+		RequestType: requestType,
+	}
+}
+
+func (r *JobRequest) needsVolume() bool {
+	return r.Job.Formation.Release.Processes[r.Type].Data
+}
+
+func (r *JobRequest) Clone() *JobRequest {
+	return &JobRequest{
+		Job:         r.Job.Clone(),
+		RequestType: r.RequestType,
+		attempts:    r.attempts,
+	}
+}
```

Comment:
The reason why both `JobRequest` and `Job` need to be copied should be documented